scrapy_poet/page_input_providers.py

```diff
@@ -329,12 +331,13 @@ def update_cache(self, request: Request, mapping: Dict[Type, Any]) -> None:
     def get_from_cache(self, request: Request, cls: Callable) -> Optional[Any]:
         return self._cached_instances.get(request, {}).get(cls)
 
-    async def __call__(
+    @inlineCallbacks
+    def __call__(
```

Comment:
What problem was happening before this change?

Reply:
It was when using `--reactor=asyncio` with pytest which fails the `asyncio` and `asyncio-min` toxenvs.

Specifically, it fails for all tests which uses `ItemProvider`:

```builtins.RuntimeError: Task got bad yield: <Deferred at 0x7f32afef6410>```

Reply:
This sounds like https://docs.scrapy.org/en/latest/topics/asyncio.html#awaiting-on-deferreds (in which case wrapping the deferred into a Future using `maybe_deferred_to_future()` should be enough).

Reply:
aha, that works! Thanks! 1bc3ee3 I was trying to use `@inlineCallbacks` similar to how the `Injector` is being set up.

Reply:
Ohh no there's an issue though where it gets stuck when running the tests on deadlock.

Do you know in particular if `maybe_deferred_to_future()` raises a `RecursionError`?

Reference: https://github.com/scrapinghub/scrapy-poet/pull/129/files#diff-a38b072945ddeae120d2462383b70ddc0f40967b53c849c49d6a3cbdf1917f53R368

Reply:
Hmm no, I have no idea about this

Reply:
I'll look into some solutions for this tomorrow. Otherwise, perhaps we can revert and put on hold the change in 1bc3ee3 for the future since it was working without problems before. In this way, it wouldn't block the upcoming release.

I think the main problem is that `asyncio.Future` doesn't increase the recursion depth which leads to the `RecursionError` that we're checking here. Another solution is to create a new check other than `RecursionError` as a proxy for detecting deadlocks but this would take more time to implement and test.

Reply:
Learned the hardway when addressing [https://github.com/scrapinghub/scrapy-poet/pull/129#discussion_r1102084934](https://github.com/scrapinghub/scrapy-poet/pull/129#discussion_r1102084934) that we really need to go into this direction. Fortunately, I've implemented a solution to solve the problem of detecting deadlocks when using asyncio's `Future` instead of Twisted's `Deferred`. eaf0d40